Partition

abh2k opened the pull request:
Fixes  #11058 

 <!-- BEGIN RELEASE NOTES -->
* functions
  *  nT function should provide faster partition counting for integers
  * nT always returns Integer for integer input
<!-- END RELEASE NOTES -->

abh2k committed:
Changes _nT function

sylee957 commented:
I don't think that this is related to the solvers module.

abhinav28071999 commented:
Oh yes, i will change it .

smichr commented:
Please post timings. This does not appear to use the referenced work in the issue.

abhinav28071999 commented:
I have used the NrParts function as mentioned in the referenced work and used it in _nT function in sympy. I have yet to implement the P function from the referenced work. 
And what is timings?

smichr commented:
> used the NrParts function

This doesn't look like `NrParts` from [here](https://github.com/klocey/partitions/blob/master/pypartitions/pypartitions.py). Am I missing something?

abhinav28071999 commented:
I found a mistake. I have not completely implemented NrParts. I will complete it as soon as possible.

abhinav28071999 commented:
@smichr Hey are there any more changes to be done in this pr.

jksuom commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1858,12 +1858,50 @@ def stirling(n, k, d=None, kind=2, signed=False):
+            for i in range(2,k1 + 1):
+                for m  in range(i + 1,q1 - i + 1 + 1):
+                    p[m] = p[m] + p[m - i]
+                    numparts = p[q1 - k1 + 1]
```
It looks like this is indented too far. It should not be necessary to assign repeatedly to `numparts` inside the double loop.

jksuom commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1858,12 +1858,50 @@ def stirling(n, k, d=None, kind=2, signed=False):
+            args = [n - k]
+    numparts = 0
+    if len(args) == 1:
+        n = args[0]
```
This is essentially the same algorithm as the one implemented in `partition`:
https://github.com/sympy/sympy/blob/4cdcbd914d66733eee1de77d592bbad3d693b049/sympy/functions/combinatorial/numbers.py#L1351-L1366

Would it be possible to use that here?

abhinav28071999 replied:
Hey please tell if this will work ,I will replace everything with 
numparts = partition(n)
when len(args) == 1

smichr replied:
What is `k` when there is 1 arg? There is no need to remove the args that are there. You can choose to handle the different cases for k and n as you see fit, however.

abhinav28071999 commented:
@jksuom  Hey, i have made the changes, please review this pr.

jksuom commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1858,12 +1858,37 @@ def stirling(n, k, d=None, kind=2, signed=False):
 
 @cacheit
-def _nT(n, k):
+def _nT(*args):
```
It would be possible to omit some repeated statements like `n = args[0]` if the signature were `_nT(n, k=None)`. That would take care of both 1 and 2 argument cases.

jksuom commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1858,36 +1858,26 @@ def stirling(n, k, d=None, kind=2, signed=False):
-            args = [n - k]
+    if k > n:
+        return 0
+    if k >= n/2.0:
```
Could this be `2*k >= n`?

abhinav28071999 replied:
yes i will change it

smichr commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1858,12 +1858,37 @@ def stirling(n, k, d=None, kind=2, signed=False):
+        elif n < k or k == 0:
+            numparts = 0
+        else:
+            q1 = int(n)
```
What is the purpose of the temporary variable here? If you eliminate this and keep the n, k arguments and assume they are ints already (since this is a private routine and the calling routine already handled that conversion) you should be able to reduce the code to half the number of current lines...or better.

abhinav28071999 replied:
Okay i will eliminate q1 and  k1

abhinav28071999 commented:
@smichr  i have made the changes please review this pr

smichr commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1858,12 +1858,27 @@ def stirling(n, k, d=None, kind=2, signed=False):
-    if k == 0:
-        return 1 if k == n else 0
-    return sum(_nT(n - k, j) for j in range(min(k, n - k) + 1))
+    if k is None:
```
eliminate this since the calling routine knows the value of k and no user should be calling this directly unless they have sanitized their data ... and know what k is ;-)

smichr commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1858,12 +1858,27 @@ def stirling(n, k, d=None, kind=2, signed=False):
+    if 2*k >= n:
+        return partition(n - k)
+    numparts = 0
+    if n == k or k == 1:
```
Why delay -- just return 1 as above where you returned 0

smichr replied:
By moving this before the next test we avoid a function call when `n ==k`.

smichr commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1858,12 +1858,27 @@ def stirling(n, k, d=None, kind=2, signed=False):
+    numparts = 0
+    if n == k or k == 1:
+        numparts = 1
+    elif n < k or k == 0:
```
Why check for `k > n` again?

How about putting this `k==0` test with the first `k > n` test?

smichr commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1858,12 +1858,27 @@ def stirling(n, k, d=None, kind=2, signed=False):
+        numparts = 0
+    else:
+        p = [1]*n
+        for i in range(2, k + 1):
```
Would this work?
```python
M = n
for i in range(2, k + 1):
    for m in range(i + 1, M):
        p[m] += p[m - i]
    M -= 1
return p[M]
```

smichr commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1858,12 +1858,27 @@ def stirling(n, k, d=None, kind=2, signed=False):
+    if k > n:
+        return 0
+    if 2*k >= n:
+        return partition(n - k)
```
I think @jksuom mentioned `_partition` so you will want to call this as `partition._partition(n - k)`

asmeurer commented:
The issue in the PR description should be "fixes #11058" with the word "fixes" so that GitHub automatically closes the issue when this is merged. It also should be separate from the release notes. I have updated it.

smichr commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1861,9 +1861,19 @@ def stirling(n, k, d=None, kind=2, signed=False):
+        return 1
+    if 2*k >= n:
+        return partition._partition(n - k)
+    M = n
```
No need for an else since the 3 cases above never get here.

smichr replied:
Somehow we need to reference that this nugget was modified from [here](https://github.com/klocey/partitions/blob/master/pypartitions/pypartitions.py).

smichr replied:
Rather than working with `n*k` we can work with (n-k)*k` with this, I believe:
```python
    M = n - k
    p = [1]*M
    for i in range(2, k + 1):
        for j in range(i + 1, M):
            p[j] += p[j - i]
        M -= 1
    # if p[0] were appended to the end of p then the last
    # k values of p are the nT(n, j) values for 0 < j < k in reverse order
    # p[-1] = nT(n, 1), p[-2] = nT(n, 2), etc.... Instead of putting the 1
    # from p[0] there, however, it is simply added to the sum below.
    return 1 + sum(p[1 - k:]) # ok for 1 < k <= n//2
```

smichr commented:
I think this is looking a lot better. I removed the `k=None` case since this is a private function for nT and nT will provide the value of `k`.

smichr commented:
Noting this here so it doesn't get lost: Somehow we need to reference that the code nugget for the `_nT` function was modified from [here](https://github.com/klocey/partitions/blob/master/pypartitions/pypartitions.py). Do you have a sense of how best to do that, @asmeurer ? Though the code has been pruned considerably, the basic structure is what was in the reference. I think it is similar to the dynamic approach that is discussed [here](https://www.geeksforgeeks.org/count-number-of-ways-to-partition-a-set-into-k-subsets/).

abhinav28071999 commented:
@smichr  Thank you for the help. Is there anything else i need to change in this pr.

smichr commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1861,9 +1861,19 @@ def stirling(n, k, d=None, kind=2, signed=False):
-    return sum(_nT(n - k, j) for j in range(min(k, n - k) + 1))
+    if k > n or k == 0:
+        return 0
+    if n == k or k == 1:
```
I believe these two lines should appear first to handle the `n==k==0` case (in case this is called directly). A test to add to the `test_numbers.py` file in the function that tests `partition` would be
```python
assert [_nT(i,j) for i in range(5) for j in range(i + 2)] == [1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 2, 1, 1, 0]
```
(That should fail in your current PR but pass once those two lines are move.)

I don't think a reference is needed after all since this is just the dynamic programming approach for this problem and is given [here](https://codereview.stackexchange.com/questions/171551/finding-number-of-integer-partitions), too.

smichr commented:
This needs a test and then part 1 of that 3 year old issue is checked off. I modified the OP so it doesn't close the issue since only half is done. Thanks for taking this up.

smichr commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1861,10 +1861,12 @@ def stirling(n, k, d=None, kind=2, signed=False):
         return 1
     if k > n or k == 0:
         return 0
+    if k == 1:
```
What values of `n` and `k` will require the 1 to be returned in two different places? i.e. what motivated this change?

abhinav28071999 commented:
@smichr   I have a question. You have made some changes to this remote branch. Now what is the command to update these changes to my local partition branch.

abhinav28071999 commented:
@smichr  when n=0 and k=1 then _nT should be equal to 0. So it was necessary to not  (n==k or k==1)
so i moved k==1 at last to resolve this .

oscarbenjamin commented:
> @smichr I have a question. You have made some changes to this remote branch. Now what is the command to update these changes to my local partition branch.

```
$ git status # Check you don't have any local changes since last commit
$ git checkout partition
$ git pull
```

abhinav28071999 commented:
@smichr  please review this i added the test case.

smichr commented:
> when n=0 and k=1 then _nT should be equal to 0

What is your reference for this? Currently in master:
```
>>> partition(0)
1
```
There is one 1 to take 0 items, not 0 ways, right? In contrast, there is no way to take 1 item in two parts.

abhinav28071999 commented:
https://en.wikipedia.org/wiki/Partition_(number_theory)#Restricted_part_size_or_number_of_parts
There is a recurrence function when n=0and k=0   it gives 1 but when i solve for n=0 and k = 1 it gives 0.

abhinav28071999 commented:
partition(0) = 1 because there is 1 way to write 0 as sum of positive numbers.It is an empty set so partition(0) = 1.
But _nT(0,1) means of partition of 0 with size=1, but empty set has size 0 so there is no was to divide 1 as sum of positive integers with size 1.

abhinav28071999 commented:
Moreover the assert case u gave assert [_nT(i,j) for i in range(5) for j in range(i + 2)] == [
    1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 2, 1, 1, 0] there also it says _nT(0,1)=0

smichr commented:
hmmm....so maybe an issue should be raised and other functions updated. But see also the code I cited:
q is our n and n is our k
```python
        if q == n or n == 1:
            numparts = 1
        elif q < n or n == 0:
            numparts = 0
```

smichr commented:
I'll be gone for the next week -- might be able to check in from time to time, else @jksuom might be available.

abhinav28071999 commented:
@smichr  The code you cited is correct. There in the code at the starting it checks 
if n >= q/2.0:
            args = [q-n]
as this is true for q=0 and n=1, it will give true and our new args will become args= [ -1 ]  now length or args is 1 now so we will evaluate p(q) which is p(-1)=0

smichr commented:
Ah, thanks for taking the time to explain. So if the 2k test came first then the old form would have been fine. If that were done then half of the time those additional tests would not have to be done.

abhinav28071999 commented:
@smichr  i have interchanged the test case. Now it is working fine. Are there any more improvements
in his pr.

jksuom commented on sympy/functions/combinatorial/tests/test_comb_numbers.py:
```diff
@@ -455,6 +456,8 @@ def test_partition():
     assert partition(p).is_positive
     assert partition(x).subs(x, 7) == 15
     assert partition(y).subs(y, 8) == 22
+    assert [nT(i, j) for i in range(5) for j in range(i + 2)] == [
```
Is this valid also for `_nT`?

abhinav28071999 replied:
Yes, nT Function calls _nT when i and j is integer.

jksuom replied:
I seem to get a different result:
```
In [2]: [_nT(i, j) for i in range(5) for j in range(i + 2)]
Out[2]: [0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 2, 1, 1, 0]
```
Perhaps this should test the private `_nT` instead of `nT`.

abhinav28071999 replied:
@jksuom  I have changed he code little bit. Now it should give correct result. Please review this.

abhinav28071999 commented:
> Ah, thanks for taking the time to explain. So if the 2k test came first then the old form would have been fine. If that were done then half of the time those additional tests would not have to be done.

If we move the 2k test atfirst then for case _nT(0,0) will give 0 which is not correct

jksuom commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1923,6 +1935,21 @@ def nT(n, k=None):
+    True
+
+
+    Another way of stating the above is that the number of partitions
```
Is this text repeated for some particular purpose?

abhinav28071999 replied:
I will change it.

abhinav28071999 replied:
@jksuom  I removed those extra text.

smichr commented:
If you move the 2k test first it should give partition._partition(0) --> 1, shouldn't it?

abhinav28071999 commented:
@smichr if we move the 2k test case first then _nT(0,0) will return 0.

jksuom commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1923,6 +1933,13 @@ def nT(n, k=None):
 
+    Another way of stating the above is that the number of partitions
+    of `n` identical items into `k` or fewer parts is just the number
+    of partititions of `n + k` identical items into `k` parts,
```
This comment (minus "Another way...") could also be used to justify the `2*k >= n` test. Then we have `k >= n - k`, so *all* partitions of `n - k` have `k` or fewer parts.

abhinav28071999 replied:
I changed it. Are there any more changes that i need to make.

abhinav28071999 replied:
@smichr  @jksuom  is there anything more to change in this pr.

smichr commented:
Why are you calling partition instead if partition._partition?

abhinav28071999 commented:
@smichr  when we call _nT(0,1 ) then previously it was calling partition._partition(-1) but _partition doesn't check for negative integers and it was returning 1 but now  partition(-1 ) return 0 as it checks for negativity.

smichr commented:
```
if k>n return 0
if k == n return 1
if n==0 return 0
if k ==2 return n//2
d = n-k
if d in (1,2) return d
if k>=d return partition._partition(d)
p=[1]*n
...
```

abhinav28071999 commented:
@smichr  i made the changes you suggested .

jksuom commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1861,12 +1861,19 @@ def stirling(n, k, d=None, kind=2, signed=False):
-    if k == 1:
+    if k == n or k == 1:
         return 1
+    if k==0:
```
This could probably be `if k <= 0`.

jksuom commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1861,12 +1861,19 @@ def stirling(n, k, d=None, kind=2, signed=False):
+    if k > n:
         return 0
-    if k == 1:
+    if k == n or k == 1:
```
How should `k = n < 0` be handled?

jksuom replied:
No problem now as `k >= 0`.

smichr replied:
n < 0, by definition of  counting sets of nonnegative integers, would be 0, wouldn't it?

smichr replied:
n < 0, by definition of  counting sets of nonnegative integers, would be 0, wouldn't it?

jksuom commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1861,9 +1861,26 @@ def stirling(n, k, d=None, kind=2, signed=False):
-    if k == 0:
-        return 1 if k == n else 0
-    return sum(_nT(n - k, j) for j in range(min(k, n - k) + 1))
+    if k > n:
```
This could be `if k < 0 or k > n`. Then `if k == 0` would suffice later.

abhinav28071999 commented:
@jksuom  i made the changes that you suggested.

jksuom commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1861,9 +1861,26 @@ def stirling(n, k, d=None, kind=2, signed=False):
 def _nT(n, k):
     """Return the partitions of ``n`` items into ``k`` parts. This
     is used by ``nT`` for the case when ``n`` is an integer."""
+    if k > n or k < 0 or n < 0:
```
If `k > n` is false then `k <= n`. If also `k < 0` is false, then `k >= 0`. In that case, `n >= 0` by transitivity, so the third condition `n < 0` will always be false and could be omitted.

abhinav28071999 replied:
Yeah, i will omit it.

jksuom commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1861,9 +1861,26 @@ def stirling(n, k, d=None, kind=2, signed=False):
+    if k == 2:
+        return n//2
+    d = n - k
+    if d in (1, 2):
```
Could this be more efficient that `if d <= 2`? We know that `k < n` and therefore `d > 0`.

abhinav28071999 replied:
I think its already there -> that is if d<=2, means d is in (1,2) and its already implemented.

jksuom replied:
The question is, which one of the two equivalent tests is more efficient, `d in (1, 2)` or `d <= 2`.

abhinav28071999 replied:
Okay, i think d<=2 is more efficient. I will change it and add more test cases.

abhinav28071999 replied:
@jksuom  @smichr  I have made the changes. Are  there any more changes in this pr to be done.

smichr replied:
Could k < 0 be n*k < 0?

abhinav28071999 replied:
@smichr  if i change k<0 to n*k<0 it will give wrong answer. e.g  when n=k=-2, then it will pass he first  condition ( e.g k>n or n*k<0)  then in 2nd condition since n=k it will return 1.

jksuom replied:
The present conditions look pretty good to me. The main part of the algorithm (starting with `M = n`) presupposes that `0 < k <= n`. The first lines of the code must handle the remaining cases.

- `k <= n` is easy to guarantee:
    - if `k > n`, return 0.

- `k > 0` is split into two parts:
    -  if `k < 0`, return 0
    -  if `k == 0`, return 1 if `n == 0` else 0
       (The `n == 0` case is currently covered by the `k == n` test.)

The other commands are optimisations taking care of special cases (saving time but not necessary).

abhinav28071999 commented:
@smichr @jksuom  can this pr be merged, if there are no more changes to be done.

jksuom commented:
I would like to see a comment explaining `partition._partition(d)`.

abhinav28071999 commented:
@jksuom  i have added the comments.

jksuom commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1860,10 +1860,30 @@ def stirling(n, k, d=None, kind=2, signed=False):
-    is used by ``nT`` for the case when ``n`` is an integer."""
+    is used by ``nT`` for the case when ``n`` is an integer.
+    The number of partitions of n into k parts is equal to the
+    number of partitions of n-k, if n >= int(q/2) so to increase
```
I think the inequality should be "k >= n - k" instead of "n >= int(q/2)". This is an implementation detail that need not be in the docstring but in a comment immediately before calling partition._partition(n - k). So something like this:

    # The number of partitions of ...
    # number of ...
    if  k >= d:

abhinav28071999 replied:
@jksuom  done.

jksuom commented:
I am satisfied with the code but I wonder if some of the 49 commits should be squashed.

smichr commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1861,9 +1861,29 @@ def stirling(n, k, d=None, kind=2, signed=False):
+    if k == 2:
+        return n//2
+    d = n - k
+    if d <= 2:
```
I believe the 2 can be changed to 3 with the given conditions preceding it.

abhinav28071999 replied:
@smichr  I changed it.

smichr commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1861,9 +1861,29 @@ def stirling(n, k, d=None, kind=2, signed=False):
+    #number of partitions of n-k, if k >= n - k so to increase
+    #the speed we are calling partition._partition(n - k).
+    if k >= d:
+        return partition._partition(d)
```
Is this code never called? How are we getting by with passing 1 argument when the function takes 2?

jksuom replied:
The caching mechanism provides the second argument.

jksuom replied:
The function we are calling is actually a wrapper with one argument.

abhinav28071999 commented:
@smichr  I have made the changes you asked. 
I have uploaded my Gsoc proposal on the wiki page. Please give me some suggestions to make it better.

smichr commented:
Since you did a merge it might be hard to squash commits together. What I generally do in such cases is this:

```
git merge master
git diff master > c1
git checkout master
git checkout -b wip
git apply c1
git commit -am 'your title'
git branch -D partition
git checkout -b partition
git branch -D wip
git push -f abhinav28071999 partition
```

smichr committed:
remove memoization so _nT can use partition cache

Given

nT(n, k) = Sum(nT(n-k,i), (i, 1, k))

When 2*k >= n - k then the cache will contain all the information needed
to calculate nT(n, k), e.g.

For example:

The cache for n=7 would be 1, 1, 2, 3, 5, 7, 11, 15,
representing partitions of 0 through 7. To compute nT(11, 4)
one could calculate nT(7, 1) + nT(7, 2) + nT(7, 3) + nT(7, 4)
but this is just partition(7) - (nT(7, 5) + nT(7, 6) + nT(7, 7))
and nT(7, 5) = nT(2, 5) = nT(2, 2) = partition(2); similarly,
nT(7, 6) = partition(1) and nT(7, 7) = partition(0).
Those partition values are already in the cache so the cache for n=7
contains what is needed to compute nT(11, 4).

smichr commented:
Failing tests mean I missed something...will debug in about 10 hours unless you figure it out before then.

smichr commented:
This should be ready after squashing commits. @jksuom , could you see if the docstring makes sense? The rewrite about recasting nT(n + k, k) as a sum didn't make sense to me -- or at least seemed a little obtuse. But perhaps you will feel the same about what is now written.

smichr commented on sympy/functions/combinatorial/tests/test_comb_numbers.py:
```diff
@@ -457,6 +458,14 @@ def test_partition():
+       assert _nT(10, 5) == 7
+       assert _nT(100, 98) == 2
+       assert _nT(100, 100) == 1
+       assert _nT(10, 3) == 8
```
When squashing, please add `assert [_nT(10, i) for i in range(11)] == [0, 1, 5, 8, 9, 7, 5, 3, 2, 1, 1]`

abhinav28071999 replied:
okay

smichr replied:
I squashed everything down in my `nt` branch. If you download that and delete your partition branch and replace it with nt and then force push that here, everything will update.

abhinav28071999 replied:
@smichr  the line for deleting partition is git -D partition but can you please tell what do you mean by replacing it with nt

smichr replied:
```
git checkout nt
git branch -D partition
git checkout -b partition
git branch -D nt
git push -f abhinav28071999 partition
```

abhinav28071999 replied:
@smichr  I have done it. Please check it.

jksuom commented:
I think that this comment is correct

    The partition of an integer ``n`` into ``k`` parts is
    equivalent to the sum of the partitions of ``n - k``
    into ``k`` or fewer parts:

but perhaps out of place in the docstring of `nT` as it concerned with the implementation.

smichr committed:
move docstring comment to inlines

smichr commented:
I think it was @pbrady once told me that I had spent more time on something than would likely ever be saved by anyone using the function....I am tempted to think that is the case here, too, for us. But here are some comparisons:
```
before
>>> t=time();ok=[nT(100,i) for i in range(100)];time()-t
1.4159998893737793
>>> t=time();ok=[nT(100,i) for i in range(100)];time()-t
1.3510000705718994

after
>>> t=time();ok=[nT(100,i) for i in range(100)];time()-t
0.03399991989135742
>>> t=time();ok=[nT(100,i) for i in range(100)];time()-t
0.013000011444091797
```
And now being non-recursive, one can replace that 100 in the previous with 1000 and not get a RuntimeError:
```
>>> t=time();ok=[nT(1000,i) for i in range(1000)];time()-t
10.938999891281128
>>> t=time();ok=[nT(1000,i) for i in range(1000)];time()-t
0.07299995422363281
```

One thing we did not discuss, however, is whether it is necessary to keep the cache on `_nT`. Since the function no longer is recursive, it might be better to drop that. Do either of you have a sense of what to do? Of course repeated, identical runs will be faster -- see previous example where first run took about 11 seconds and the second took 0.07 -- but I'm not sure that is a good enough reason to keep it.

ishandutta2007 commented:
For generating till N, it could be faster if you precompute the pentagonal number out in a loop.
Only you would need O(N) more memory

```
import time
start = time.time()
N = 10000
A001318 = []
for i in range(1, N):
    if i % 2 == 0:
        v = i // 2
        A001318.append(v * (3 * v + 1) // 2)
    else:
        v = i // 2 + 1
        A001318.append(v * (3 * v - 1) // 2)

A000041 = [0] * N
A000041[0] = 1
for i in range(1, N):
    v = 0
    for j in A001318:
        if j > i:
            break
        if v < 2:
            A000041[i] += A000041[i - j]
        else:
            A000041[i] -= A000041[i - j]
        v = (v + 1) % 4
print(time.time() - start)

```

> 0.4687314033508301

```
N = 10000
start = time.time()
from sympy.functions import partition
[partition(n) for n in range(N)]
print(time.time() - start)

```

> 3.1725587844848633
